boxtree/tree.py

```diff
@@ -111,9 +112,111 @@ class box_flags_enum(Enum):  # noqa
 # }}}
 
 
+# {{{ tree of boxes
+
+class TreeOfBoxes:
+    """A quad/octree tree of abstract boxes. Lightweight trees handled with
+    :mod:`numpy`, intended for mesh adaptivity.
+
+    .. attribute:: dimensions
+
+    .. attribute:: nlevels
+
+    .. attribute:: nboxes
+
+    .. attribute:: root_extent
+
+        (Scalar) extent of the root box.
+
+    .. attribute:: box_centers
+
+        dim-by-nboxes :mod:`numpy` array of the centers of the boxes.
+
+    .. attribute:: box_parent_ids
+
+        :mod:`numpy` vector of parent box ids.
+
+    .. attribute:: box_child_ids
+
+        (2**dim)-by-nboxes :mod:`numpy` array of children box ids.
+
+    .. attribute:: box_levels
+
+        :mod:`numpy` vector of box levels.
+
+    .. automethod:: __init__
+
+    .. automethod:: copy
+
+    .. automethod:: get_leaf_flags
+
+    .. automethod:: leaf_boxes
+    """
+    def __init__(
+            self, box_centers, root_extent,
+            box_parent_ids, box_child_ids, box_levels):
+        self.box_centers = box_centers
+        self.root_extent = root_extent
+        self.box_parent_ids = box_parent_ids
+        self.box_child_ids = box_child_ids
+        self.box_levels = box_levels
+
+        lows = self.box_centers[:, 0] - 0.5*self.root_extent
+        highs = lows + self.root_extent
+        self.bounding_box = [lows, highs]
+
+        self.dim = self.box_centers.shape[0]
```

Comment:
Would it be reasonable to make this a `dataclass`?

Reply:
a609966